mol_dqn/chemgraph/target_sas_eval.ipynb

```diff
@@ -0,0 +1,227 @@
+        "      allow_no_modification=hparams.allow_no_modification,\n",
+        "      allowed_ring_sizes=set(hparams.allowed_ring_sizes),\n",
+        "      allow_bonds_between_rings=hparams.allow_bonds_between_rings,\n",
+        "      max_steps=hparams.max_steps_per_episode)\n",
+        "\n",
+        "  dqn = deep_q_networks.DeepQNetwork(\n",
+        "      input_shape=(hparams.batch_size, hparams.fingerprint_length + 1),\n",
+        "      q_fn=functools.partial(\n",
+        "          deep_q_networks.multi_layer_model, hparams=hparams),\n",
+        "      optimizer=hparams.optimizer,\n",
+        "      grad_clipping=hparams.grad_clipping,\n",
+        "      num_bootstrap_heads=hparams.num_bootstrap_heads,\n",
+        "      gamma=hparams.gamma,\n",
+        "      epsilon=0.0)\n",
+        "  \n",
+        "  tf.reset_default_graph()\n",
+        "  with tf.Session() as sess:\n",
+        "    dqn.build()\n",
+        "    model_saver = tf.train.Saver(max_to_keep=hparams.max_num_checkpoints)\n",
+        "    model_saver.restore(sess, os.path.join(model_dir, 'ckpt-%i' % ckpt))\n",
+        "    environment.initialize()\n",
+        "    for step in range(hparams.max_steps_per_episode):\n",
+        "      steps_left = hparams.max_steps_per_episode - environment.num_steps_taken\n",
+        "      \n",
+        "      if hparams.num_bootstrap_heads:\n",
+        "        head = np.random.randint(hparams.num_bootstrap_heads)\n",
+        "      else:\n",
+        "        head = 0\n",
+        "      valid_actions = list(environment.get_valid_actions())\n",
+        "      observations = np.vstack(\n",
+        "        [np.append(deep_q_networks.get_fingerprint(act, hparams), steps_left) \n",
+        "         for act in valid_actions])\n",
+        "      action = valid_actions[dqn.get_action(\n",
+        "          observations, head=head, update_epsilon=0.0)]\n",
+        "      result = environment.step(action)\n",
+        "  return ckpt, result\n"
+      ],
+      "outputs": [],
+      "execution_count": null,
+      "metadata": {}
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "all_results = []\n",
+        "for i in range(31):\n",
+        "    for target in (2.5, 4.8):\n",
+        "        ckpt, result = eval(basepath %(i, target), i)\n",
+        "        ori_sas = sascorer.calculateScore(Chem.MolFromSmiles(all_molecules[i]))\n",
+        "        sas = sascorer.calculateScore(Chem.MolFromSmiles(result.state))\n",
+        "        all_results.append((i, ckpt, target, sas))"
+      ],
+      "outputs": [],
+      "execution_count": null,
+      "metadata": {
+        "scrolled": true
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "df = pd.DataFrame(all_results, columns=['index', 'ckpt', 'target_sas', 'sas'])\n",
+        "ori_sas = sum([[sascorer.calculateScore(Chem.MolFromSmiles(all_molecules[i]))] * 2 for i in range(31)], [])\n",
```

Comment:
This line doesn't make sense to me; what are you trying to do?

Reply:
I want to include the original SAS in the table. On line 156, I forgot to include `ori_sas` in the result, and I don't want to do the whole evaluation again (~1h). Here is a remedy for that.

Reply:
Yes, I understood that, but I don't understand the syntax. You are doing:

`sum([[sascorer.calculateScore(Chem.MolFromSmiles(all_molecules[i]))] * 2 for i in range(31)], [])`

which looks to me like you are multiplying the SA score by 2 (why?), also I don't understand why you are calling `sum`...

Reply:
OK, I think I understand what this is doing now.

You are computing the SA score for a molecule, then repeating it (` * 2`), so you end up with a list of lists of length 2. Then you want to flatten this list, which `sum(..., [])` does for you (although I've never seen that syntax before).

This is complicated enough that I would like you to choose another way to do this. Perhaps use a few more lines :)

Reply:
`* 2` was applied on the list.
it is actually sum([[sas] * 2 for sas in SAS], [])
[sas] * 2 gets [sas, sas], because each molecule is evaluated for 2.5 and 4.8.
sum() combines this to a single list.

Reply:
Yeah. I can rewrite it, but it is the original code I use. 

I added a new cell under this cell to explain it is actually the same as that line